### PR TITLE
【iOS】Fixes ellipsis carries background from trimmed text

### LINKED ISSUE #37926: [iOS] Ellipsis carries background from trimmed text
### Description

This issue happens on iOS only:

I tried to use nested text which has background, and could be wrapped with an ellipsis. Even the nested text doesn't show, the ellipsis has the background of nested text.

```
<Text numberOfLines={1} ellipsizeMode={'tail'}>Long text <Text style={{backgroundColor: 'red'}}>Nested text</Text></Text>
```

### React Native Version

0.70.6

### Output of `npx react-native info`

System:
    OS: macOS 12.5.1
    CPU: (10) arm64 Apple M1 Max
    Memory: 548.38 MB / 64.00 GB
    Shell: 5.8.1 - /bin/zsh
  Binaries:
    Node: 16.13.0 - ~/.nvm/versions/node/v16.13.0/bin/node
    Yarn: 1.19.2 - ~/.nix-profile/bin/yarn
    npm: 8.1.0 - ~/.nvm/versions/node/v16.13.0/bin/npm
    Watchman: 4.9.0 - /Users/shidong.lin/.nix-profile/bin/watchman
  Managers:
    CocoaPods: 1.11.0 - /Users/shidong.lin/.nix-profile/bin/pod
  SDKs:
    iOS SDK:
      Platforms: DriverKit 22.2, iOS 16.2, macOS 13.1, tvOS 16.1, watchOS 9.1
    Android SDK:
      API Levels: 30, 31, 32
      Build Tools: 30.0.3
      System Images: android-31 | Google Play ARM 64 v8a, android-32 | Google Play ARM 64 v8a
      Android NDK: Not Found
  IDEs:
    Android Studio: 2022.2 AI-222.4459.24.2221.9862592
    Xcode: 14.2/14C18 - /usr/bin/xcodebuild
  Languages:
    Java: 11.0.11 - /Users/shidong.lin/.nix-profile/bin/javac
  npmPackages:
    @react-native-community/cli: Not Found
    react: 18.2.0 => 18.2.0
    react-native: https://github.com/react-native/archive/f56cf181afcd364122f94ac8609f9b3c7b0b1afb.tar.gz => 0.70.6
    react-native-macos: Not Found
  npmGlobalPackages:
    *react-native*: Not Found

### Steps to reproduce

```
<Text numberOfLines={1}>Long text <Text style={{backgroundColor: 'red'}}>Nested text</Text></Text>
```
Just create a long text with colored nested text, check the behavior of ellipsis.

### Snack, code example, screenshot, or link to a repository

```
<Text numberOfLines={1} ellipsizeMode={'tail'}>Long text <Text style={{backgroundColor: 'red'}}>Nested text</Text></Text>
```

### PR DESCRIPTION
**Pull request by zhongwuzw**
## Summary:

Fixes #37926. 

## Changelog:

[IOS] [FIXED] - Fixes ellipsis carries background from trimmed text

## Test Plan:

Demo code: 
``` 
<Text numberOfLines={1} ellipsizeMode={'tail'}>Long text <Text style={{backgroundColor: 'red'}}>Nested text</Text></Text>
```

Before fix: 
image

After fix:
image

### CODE REVIEW

**Comment by javache** [comment]
I don't think we want to mutate `_textStorage` in the draw path. Would it be possible to do so in the layout path? Also, could you verify this behaviour in the new architecture too?

**Comment by zhongwuzw** [comment]
> I don't think we want to mutate `_textStorage` in the draw path. Would it be possible to do so in the layout path? Also, could you verify this behaviour in the new architecture too?

@javache Thanks for the review, I have no idea how to do it in the layout path, my thought is we can mutate the textStorage before drawing and restore the textStorage which we mutated after drawing. 

And I tested in new arch and it has the same issue.

**Comment by javache** [comment]
I would look at making this change in [RCTTextShadowView](https://github.com/facebook/react-native/blob/eddefec5f8dc58a1ecf66944e36d8c780fac7064/packages/react-native/Libraries/Text/Text/RCTTextShadowView.mm#L203) and [RCTTextLayoutManager](https://github.com/facebook/react-native/blob/eddefec5f8dc58a1ecf66944e36d8c780fac7064/packages/react-native/ReactCommon/react/renderer/textlayoutmanager/platform/ios/react/renderer/textlayoutmanager/RCTTextLayoutManager.mm#L323) (for the new architecture)

**Comment by zhongwuzw** [comment]
@javache Friendly ping, thank you for any feedback :)

**Review by javache — CHANGES_REQUESTED** [review]
Please see my previous feedback about moving this logic on the background thread by using RCTTextShadowView/RCTTextLayoutManager. Please also ensure this is compatible with the new architecture.

File: packages/react-native/ReactCommon/react/renderer/components/text/ParagraphLayoutManager.cpp [hunk 1]
```diff
@@ -32,12 +32,14 @@ TextMeasurement ParagraphLayoutManager::measure(
...
   } else {
+    hostTextStorage_ = textLayoutManager_->getHostTextStorage(
+        attributedString, paragraphAttributes, layoutConstraints);
     return textLayoutManager_->measure(
         AttributedStringBox(attributedString),
         paragraphAttributes,
         layoutContext,
         layoutConstraints,
-        nullptr);
+        hostTextStorage_);
```
Comment (zhongwuzw) [inline packages/react-native/ReactCommon/react/renderer/components/text/ParagraphLayoutManager.cpp:42]: Seems Fabric uses different `textStorage`  in measure and draw path. We should share it.

**Comment by zhongwuzw** [comment]
@javache Seems we can't put it into background thread In new arch, because we use separate textstorage for measure and draw path, because shared of textstorage was removed in https://github.com/facebook/react-native/pull/43914.

**Comment by zhongwuzw** [comment]
any one can help to review? cc. @sammy-SC @javache

**Comment by javache** [comment]
Thanks for investigating this. The changes look reasonable, but can we add gating here so we only execute this if numberOfLines is set?

**Comment by zhongwuzw** [comment]
@javache Done :)

**Comment by cipolleschi** [comment]
Hi @zhongwuzw, I know that you are fixing a specific issue, but there could be other style properties that we want to remove from the ellipsis.

I think that the ellispis should keep the same style of the character that precedes it, not to have weird artifacts on the screen.
what do you think?

**Comment by zhongwuzw** [comment]
@cipolleschi Hi, I just keep the same color related attributes as the character that precedes it. Other attributes like fontSize would impact the layout of the text.

**Comment by javache** [comment]
This is failing internal builds with 

```
packages/react-native/Libraries/Text/Text/RCTTextShadowView.mm:262:64: error: declaration shadows a local variable [-Werror,-Wshadow]
                                     NSTextContainer *_Nonnull textContainer,
```

**Comment by zhongwuzw** [comment]
@javache done 😀

**Pull request closed without merging**